### PR TITLE
ARROW-12607: [Website] Doc section for Dataset Java bindings

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

File: docs/source/java/dataset.rst [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+.. code-block:: Java
+
+    Scanner scanner = dataset.newScan(new ScanOptions(new String[] {"id", "name"}, 100));
+    Schema projectedSchema = scanner.schema();
+ 
+Projection
+===========
+
+User can specify projections in ScanOptions. For ``FileSystemDataset``, only column projection is allowed for now. Which means, only column names
```
Comment (kiszk) [inline docs/source/java/dataset.rst:99]: nit: now. -> now,  ??

File: docs/source/java/dataset.rst [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    String uri = "hdfs://{hdfs_host}:{port}/data/example.parquet";
+    BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+    DatasetFactory factory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(),
+        FileFormat.PARQUET, uri);
+        
+Native Memory Management
+===========
+
+To gain better performance and reducing code complexity, Java ``FileSystemDataset`` internally relys on C++ ``arrow::dataset::FileSystemDataset`` via JNI.
+As a result, All Arrow data read from ``FileSystemDataset`` is supposed to be allocated off the JVM heap. To manage this part of memory, An utility class
```
Comment (kiszk) [inline docs/source/java/dataset.rst:134]: All -> all

File: docs/source/java/dataset.rst [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    
+    String uri = "hdfs://{hdfs_host}:{port}/data/example.parquet";
+    BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
+    DatasetFactory factory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(),
+        FileFormat.PARQUET, uri);
+        
+Native Memory Management
+===========
+
+To gain better performance and reducing code complexity, Java ``FileSystemDataset`` internally relys on C++ ``arrow::dataset::FileSystemDataset`` via JNI.
```
Comment (kiszk) [inline docs/source/java/dataset.rst:133]: reducing -> reduce
?

File: docs/source/java/dataset.rst [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+As a result, All Arrow data read from ``FileSystemDataset`` is supposed to be allocated off the JVM heap. To manage this part of memory, An utility class
+``NativeMemoryPool`` is provided to users.
+
+As a basic example, by using a listenable ``NativeMemoryPool``, User can pass a listener hooking on C++ buffer allocation/deallocation:
```
Comment (kiszk) [inline docs/source/java/dataset.rst:137]: User -> user

**Comment by zhztheplayer** [comment]
Thanks @kiszk for helping checking this :)

File: docs/source/java/dataset.rst [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+===========
+Dataset
+===========
+
+.. warning::
+
+    Experimental: The Java module ``dataset`` is currently under early development. API might be changed in each release of Apache Arrow until it gets mature.
+
+Dataset is an universal layer in Apache Arrow for querying data in different formats or in different paritionning strategies. Usually the data to be queried is
```
Comment (kiszk) [inline docs/source/java/dataset.rst:26]: nit: paritionning -> partitioning

File: docs/source/java/dataset.rst [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    ScanOptions options = new ScanOptions(projection, 100);
+    
+If no projection is needed, specify an empty String array ``new String[0]`` in ScanOptions:
+
+.. code-block:: Java
+
+    String[] projection = new String[0];
+    ScanOptions options = new ScanOptions(projection, 100);
+    
+This way all column will be emitted during scanning.
```
Comment (kiszk) [inline docs/source/java/dataset.rst:114]: nit: all column -> all columns
?

**Review by kszucs — APPROVED** [review]

**Pull request closed without merging**